experimental 4player Counter / Co-Operative

cylonicboom opened the pull request:
# 4 player pvp missions

![Screenshot from 2024-07-05 20-42-26](https://github.com/fgsfdsfgs/perfect_dark/assets/16697548/9b583423-7b96-4e3c-8d82-2fcf3b657582)
![Screenshot from 2024-07-05 20-46-53](https://github.com/fgsfdsfgs/perfect_dark/assets/16697548/2a0e0ee3-ec16-44c9-b597-336d890a276e)



# known issues

- coop + anti is still janky
- dark rooms + NVG or x-ray vision may be a flashing light hazard


# goals

Want to play 4player coop / anti with a mix of coop / anti team characters

No recompiling of ailists. Should "just work" with vanilla levels and well-formed mods.

Load mp profile for coop / anti characters for controller / hud settings / "leveling"

# non-goals

- Custom characters as coop characters (for now)
- Fixing non-breaking n64 bugs beyond what is required
- adding coop / anti support for levels not built for it
- adding anti support to mp ailists commands that don't use it in retail until modders who actually make levels also weigh in on how they should work 

# background

In Vanilla PD,  'coopness' or 'antiness' of a PD session is controlled by setting `g_Vars.coopplayernum` and `g_Vars.antiplayernum` This is used to determine if coop or anti is running, and what player index to use. Also, coop / anti logic is mutually exclusive.

# approach

Porting PD to 4 player local coop/anti requires is set via `g_MissionConfig.isteam` that allows players to join a solo mission as a team of coop or anti characters.


`g_MissionConfig.isteam` is set to `true` when the mission is a team mission.
`g_Vars.playerroles` is used to track player role constants against player slots. This is used by `playermgrAllocatePlayersFromRoles` to initialize the player slots and various control variables. 

`g_Vars.currentplayer` is the current player struct.
`g_Vars.players` : player structs for all players.
`g_Vars.currentplayernum` is the current player number

`g_Vars.coopplayers` : player structs for coop players. The bond slot is set to 0 as well as the anti players.
`g_Vars.currentcoopnum` is the current coop player number. On each frame the coop player is set to the next coop player struct that exists.
`g_Vars.coop`: current coop player struct

`g_Vars.currentantinum` is the current anti player number. On each frame the anti player is set to the next anti player struct that exists
`g_Vars.antiplayers` : player structs for anti players. The anti index is set to 0 as well as the coop players.
`g_Vars.anti`: current anti player struct


In order to gaurentee that ailist commands that are intending to reference `CHR_ANTI`, `CHR_COOP`, and `CHR_P1P2`/`CHR_P1P2O`AIlist commands referencing `CHR_COOP` or `CHR_P1P2` will be treated as any or all coop / ally chaaracters.


# limitations

- scripts targeting `CHR_COOP` or `CHR_P1P2` will need to be treated as any or all ally characters. 
This means that mission-related hud messages intended for one coop character will be sent to them all.

# acknowledgements

- borrowed fixes from @iamgreaser's 4player anti effort: merging the tiktok mode hud settings into 4p, correctly spawning 4 players into a mission
-  @fgsfdsfgs  / Ryan Dwyer for the pc port / decomp but specifically reviewing the scheduling logic and talking me down from attempting any kind of live ailist reinterpretation.

iamgreaser committed:
Use vertical-split versions of menus in 3+-player main missions

These work fine in the 2x2 split arrangement.

iamgreaser committed:
patch written by @iamgreaser

... but manually typed in

see https://github.com/fgsfdsfgs/perfect_dark/issues/383

cylonicboom committed:
treat `g_Vars.coop` and `g_Vars.anti` like `g_Vars.currentplayer`

This allows all 4 players to be detected and targeted by enemy guards in
coop

cylonicboom committed:
objective check: check all coop players for required objective item

cylonicboom committed:
p1p2_opposite is treated as currentcoopnum

iamgreaser commented:
I'm having a look at the ailists for a hopefully cleaner solution and it's starting to look a lot more feasible than I thought it would.

Having said that, having a list of all `anti` and `coop` players is a nice idea. Not sure how tempted I am to borrow that idea, but I do intend to have something working to the point where you'd have co-op and counter-op stuff happening in the same game mode.

Something you'll want to consider is the `CHR_P1P2` reference and the `chr_toggle_p1p2(CHR_SELF)` command. If you cycle through all protagonist players then I suspect that would tackle the most general cases. But there will definitely be plenty of pain especially pertaining to objectives being failed in situations where they should not be, and also stuff appearing and disappearing while visible to some of the extra co-op players.

cylonicboom committed:
4pcoop changes

use currentcoopplayernum instead of coopplayernum

p1p2_opposite: attempt to resolve a new player by subtractng the
coopplayerindex from the max number of human coop buddies

NeonNyan commented:
Cycling  the "current coop" player and basing the p1p2 toggle calculation based on MAX_PLAYERS and the current coop number number does make most things work. Like @fgsfdsfgs suggested CHR_COOP / CHR_P1P2_OPPOSITE references in the `if_chr_.*(chr,[... ,] label)` should check for all the other coop characters (when p1p2_opposite isn't bond) so the ailist can make the appropriate jump. Right now non-bond characters need to tap the use button repeatedly for some ailist scripted objects like lifts and objective items because of the coop cycling. Also the surveillance cameras take a long time to notice non-bond players.

cylonicboom committed:
add orgfile tracking 4pcoopanti ailist progress

cylonicboom committed:
if_chr_dead: handle anti/coop

cylonicboom committed:
update notes

cylonicboom committed:
`aiChrIfDead` respects `g_Vars.playerorder`

cylonicboom committed:
`aiIfChrDeathAnimationFinished` now supports 4pcoop and 4panti

untested with anti

cylonicboom committed:
helper function

cylonicboom committed:
update coopanti notes

cylonicboom committed:
update notes

cylonicboom committed:
cctvtick will enumerate over all players and break if it finds one bond
or coop player

cylonicboom committed:
setchrflags

cylonicboom committed:
aiGiveObjectToChr

cylonicboom committed:
propobj checks if any coop player is using it

cylonicboom committed:
some helper functions for checking coop / antiness

cylonicboom committed:
endscreen: setNumPlayers(4)

cylonicboom committed:
update notes

cylonicboom committed:
fixup: 4pcoop: cleanup check

cylonicboom committed:
4pcoop: spawn radius fix for multiple human players

cylonicboom committed:
4pcoop: remove comment

cylonicboom committed:
4pcoop: fixup solo / coop picking for head / body determination

cylonicboom committed:
4pcoop: shorten the length of time in front of cctv before alarm trips

cylonicboom committed:
more reliable isanti check

cylonicboom committed:
increase spawn distance

cylonicboom committed:
fixup: getPlayerPool no longer static

cylonicboom committed:
cycle aiChrUnsetChrflag per-human hench

cylonicboom committed:
cycle aiIfChrActivatedObject between all human henches of that chr type

cylonicboom committed:
refactor helper functions

cylonicboom committed:
cycle aiIfChrHasWeaponEquipped

cylonicboom committed:
fixes aiIfChrActivate

cylonicboom committed:
fixup aiIfChrHasWeaponEquipped

cylonicboom committed:
fixup isdead command

cylonicboom committed:
make aiIfChrDistanceToPadGreaterThan support >2 player coop

cylonicboom committed:
treat aiIfPlayerLookingAtObject as "any coop"

cylonicboom committed:
added new helper functions

cylonicboom committed:
when coop, make chrGetDistanceToTarget return the min of all coop players

cylonicboom committed:
treat aiIfCheckFovWithTarget as any

cylonicboom committed:
removeme: debug loglines

cylonicboom committed:
revert chrGetDistacneToTarget changes

cylonicboom committed:
cycle aiIfDistanceToTargetLessThan per coop player

cylonicboom committed:
handle aiIfChrHasObject per coop player correctly

cylonicboom committed:
remove debug output

cylonicboom committed:
Added new mode: Team Missions

Team mission starts as 1 player, but  mplayerisrunning=true

Added properties to `struct missionconfig`:
- `u8 isteam`

Added experimental hud render function `renderTeamMissionJoinText` when
controllers are connected during Team Mission gameplay. It's pasted
logic from the "Player: %d Press START / READY" stanza from `menuRender`

The plan is that for "in game" checks and scripts we'll continue to rely on coopplayernum / antiplayernum where possible. Pre-game menu setup flows will use `g_MissionConfig.isteam`.

TODO: set coopplayernum to -1 when currentplayernum is anti. (setCurrentPlayer) team missions only.
TODO: set antiplayernum to -1 when currentplayernum is coop. (setCurrentPlayer) team missions only.

Restored vanilla coop to retail behavior

Player 2 can be live-joined via gdb:

```gdb
define load_coop_player
  if $argc != 1
    printf "load_player: Usage: load_player <player_id>\n"
    return
  end

  # this logic was copied from lv.c:lvReset
  call playermgrAllocatePlayer($arg0)

  set $lastplayer = g_Vars.currentplayernum
  call setCurrentPlayerNum($arg0)
  set g_Vars.currentplayer->usedowntime = 0
  set g_Vars.currentplayer->invdowntime = g_Vars.currentplayer->usedowntime

  call menuReset()
  call amReset()
  call invReset()
  call bgunReset()
  call playerLoadDefaults()
  call playerReset()
  call playerSpawn()
  call bheadReset()

  call setCurrentPlayerNum($lastplayer)
end

define load_anti_player
  if $argc != 1
    printf "load_player: Usage: load_player <player_id>\n"
    return
  end

  # this logic was copied from lv.c:lvReset
  call playermgrAllocatePlayer($arg0)

  set $lastplayer = g_Vars.currentplayernum
  call setCurrentPlayerNum($arg0)
  set g_Vars.antiplayers[$arg0] = g_Vars.currentplayer
  set g_Vars.currentplayer->usedowntime = 0
  set g_Vars.currentplayer->invdowntime = g_Vars.currentplayer->usedowntime

  call menuReset()
  call amReset()
  call invReset()
  call bgunReset()
  call playerLoadDefaults()
  call playerReset()
  call playerSpawn()
  call bheadReset()

  call setCurrentPlayerNum($lastplayer)
end

setNumPlayers(2)
load_coop_player 1 # or load_anti_player 1
continue
```

cylonicboom committed:
Implement player assignment menu

counter-op selection is currently broken

cylonicboom committed:
supports coop vs anti

cylonicboom committed:
some more fixes to coop / anti scheduling

cylonicboom committed:
non-contigious controller assignments work

block team mdoe if < 1 coop or anti assigned

coop ailists seem to be broken again (guards cant detect)

coop / anti coex is wonky w/ crosshair targetting

cylonicboom committed:
fix windows build issue?

cylonicboom committed:
windows build issue

cylonicboom committed:
windows build issue

cylonicboom committed:
windows bld issue

cylonicboom committed:
windows build issue

cylonicboom committed:
windows build issue

cylonicboom committed:
windows build issue

cylonicboom committed:
windows build issue

cylonicboom committed:
windows build issues

cylonicboom committed:
fixup playerroles

cylonicboom committed:
windows build issue

cylonicboom committed:
windows build issue

cylonicboom committed:
windows build issue

cylonicboom committed:
Renamed "Team Missions" -> Quad-Operative

...per waifu's suggestion

cylonicboom committed:
fix p1p2 scheduling issues

JacksBestGaming commented:
I have no clue who came up with 'Build Your Perfect Team", but I love it so much. This is not only extremely cool on a technical level, but on a creative one as well.

Unradelic commented:
![image](https://github.com/user-attachments/assets/e8f5e76a-adc5-4c63-b91a-0af68507b773)

Happened while fiddling with the menu options, going back and forth, then crashed :P

cylonicboom committed:
really fix p1p2 scheduling issues

revert: allyplayers was a bad idea

NeonNyan commented:
> ![image](https://private-user-images.githubusercontent.com/20099874/348471718-e8f5e76a-adc5-4c63-b91a-0af68507b773.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjQwMDE2MDIsIm5iZiI6MTcyNDAwMTMwMiwicGF0aCI6Ii8yMDA5OTg3NC8zNDg0NzE3MTgtZThmNWU3NmEtYWRjNS00YzYzLWI5MWEtMGFmNjg1MDdiNzczLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MTglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODE4VDE3MTUwMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWIzMzEyNzMyOGNiYjY3MTg3ZTNmNzU4ZGE0NDAwOTE2ZmFiN2MxOTNjMjRlMzk5ODNmZThiMTg2NzE4ZjM4MmMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.JVjnNrUMPDEPXwhUuu730pObtSFLACg__jx_8AR2c4c)
> 
> Happened while fiddling with the menu options, going back and forth, then crashed :P

Pretty sure I've seen this while trying to get controllers setup.

cylonicboom committed:
fix player allocation

cylonicboom committed:
use PLAYERCOUNT() instead of MAX_PLAYERS

cylonicboom committed:
don't set `g_Vars.mplayerisrunning = true` in menus

fixes menu crashes?

NeonNyan commented:
> ![image](https://private-user-images.githubusercontent.com/20099874/348471718-e8f5e76a-adc5-4c63-b91a-0af68507b773.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjQwMDE2MDIsIm5iZiI6MTcyNDAwMTMwMiwicGF0aCI6Ii8yMDA5OTg3NC8zNDg0NzE3MTgtZThmNWU3NmEtYWRjNS00YzYzLWI5MWEtMGFmNjg1MDdiNzczLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MTglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODE4VDE3MTUwMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWIzMzEyNzMyOGNiYjY3MTg3ZTNmNzU4ZGE0NDAwOTE2ZmFiN2MxOTNjMjRlMzk5ODNmZThiMTg2NzE4ZjM4MmMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.JVjnNrUMPDEPXwhUuu730pObtSFLACg__jx_8AR2c4c)
> 
> Happened while fiddling with the menu options, going back and forth, then crashed :P

https://github.com/fgsfdsfgs/perfect_dark/pull/389/commits/cf75ba31b05d9dbfc45b082f80f0af448b4d585b should fix it if it's the same issue I encountered.

NeonNyan commented:
[ ] BUG: vanilla allocation flow disabled and this breaks vanilla campaign flows

cylonicboom committed:
fixup: aiIfChrDeathAnimationFinished no longer incorrectly marks coop as dead

cylonicboom committed:
fixup: use seperate continue button handlers for team / coop modes

cylonicboom committed:
coop support for aiRevokeControl

cylonicboom committed:
teleportals kind of work

cylonicboom committed:
fixup: this will help us move around the ship quicker

MonsMarineris commented:
![Screenshot 2024-09-30 154934](https://github.com/user-attachments/assets/db0a60a3-832e-442d-a091-dabf644756cc)
I am unable to build this for some reason :(
I do admit I have no fuggin clue what I'm doing. But still, I was so excited to try this out.

cylonicboom commented on src/game/chraicommands.c:
```diff
@@ -5125,22 +5355,45 @@ bool aiRevokeControl(void)
-		bgunSetSightVisible(GUNSIGHTREASON_NOCONTROL, false);
-		bgunSetGunAmmoVisible(GUNAMMOREASON_NOCONTROL, false);
+		if (isChrPropCoop(chr->prop)) {
+			struct playerpool** playerpool = getPlayerPool(CHR_COOP);
```
wtf I have no idea how this built and worked on Linux. I'll take a look when I'm on my computer next. Unblocks https://github.com/fgsfdsfgs/perfect_dark/pull/389#issuecomment-2384121638

NeonNyan commented:
> ![Screenshot 2024-09-30 154934](https://github.com/user-attachments/assets/db0a60a3-832e-442d-a091-dabf644756cc)
> 
> I am unable to build this for some reason :(
> 
> I do admit I have no fuggin clue what I'm doing. But still, I was so excited to try this out.

I broke windows builds with some stupid code that I'm not sure how it built on Linux. I'll take a look when I'm on my computer next.

NeonNyan commented:
I'm closing this PR and taking it private until it's in a more stable state.

MonsMarineris commented:
Thanks for having a look at it. I wish I had any sort of knowledge to help. I'm trying to learn.
